src/rust/src/x509/verify.rs

```diff
@@ -43,6 +47,134 @@ impl CryptoOps for PyCryptoOps {
     }
 }
 
+struct FixedPolicy<'a>(Policy<'a, PyCryptoOps>);
+
+/// This enum exists solely to provide heterogeneously typed ownership for `OwnedPolicy`.
+enum SubjectOwner {
+    // TODO: Switch this to `Py<PyString>` once Pyo3's `to_str()` preserves a
+    // lifetime relationship between an a `PyString` and its borrowed `&str`
+    // reference in all limited API builds. PyO3 can't currently do that in
+    // older limited API builds because it needs `PyUnicode_AsUTF8AndSize` to do
+    // so, which was only stabilized with 3.10.
+    DNSName((pyo3::Py<pyo3::PyAny>, String)),
+    IPAddress((pyo3::Py<pyo3::PyAny>, pyo3::Py<pyo3::types::PyBytes>)),
+}
+
+self_cell::self_cell!(
+    struct OwnedPolicy {
+        owner: SubjectOwner,
+
+        #[covariant]
+        dependent: FixedPolicy,
+    }
+);
+
+#[pyo3::pyclass(
+    name = "ServerVerifier",
+    module = "cryptography.hazmat.bindings._rust.x509"
+)]
+struct PyServerVerifier(OwnedPolicy);
+
+impl PyServerVerifier {
+    fn as_policy(&self) -> &Policy<'_, PyCryptoOps> {
+        &self.0.borrow_dependent().0
+    }
+}
+
+#[pyo3::pymethods]
+impl PyServerVerifier {
+    #[getter]
+    fn subject<'p>(&'p self, py: pyo3::Python<'p>) -> pyo3::PyResult<Option<&'p pyo3::PyAny>> {
+        match self.0.borrow_owner() {
+            SubjectOwner::DNSName((subject, _)) => Ok(Some(subject.as_ref(py))),
+            SubjectOwner::IPAddress((subject, _)) => Ok(Some(subject.as_ref(py))),
+        }
+    }
+
+    #[getter]
+    fn validation_time<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        datetime_to_py(py, &self.as_policy().validation_time)
+    }
+}
+
+fn build_subject_owner(
+    py: pyo3::Python<'_>,
+    subject: pyo3::Py<pyo3::PyAny>,
+) -> pyo3::PyResult<SubjectOwner> {
+    let subject = subject.as_ref(py);
+
+    let x509_general_name_module =
+        py.import(pyo3::intern!(py, "cryptography.x509.general_name"))?;
```

Comment:
We've replaced this pattern with the one seen in `types.rs`, basically never use `py.import` :-)

Reply:
[a5a88e6](https://github.com/pyca/cryptography/pull/9405/commits/a5a88e6291c7f85e8a1915d70c618ba407a5a56a)